### PR TITLE
API: Account changes to event stream

### PR DESCRIPTION
**Pull request by kfangw**
### Requirements

* Implement SSE(Sever Sent Event)

### Github Issue
#3 

### Solution
I changed two point
- server-sent-event api with go-observerble
- Adopt mux

After this change, i will add bellows
- client
- APIs for each issues

### CODE REVIEW

**Comment by kfangw** [comment]
@spikeekips @zzim2x Please review

File: lib/broker_test.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	time.Sleep(time.Second)
+
+	w.Flush()
+	w2.Flush()
+	ret1 := buffer.String()
+	ret2 := buffer2.String()
+
+	fmt.Println(ret1)
+	fmt.Println(ret2)
```
Comment (outsideris) [inline lib/broker_test.go]: These are unnecessary.

File: lib/broker.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import (
+	"io"
+)
+
+type EventType string
+
+const (
+	EventDummy  EventType = "dummy"
+	EventDummy2           = "dummy2"
+)
```
Comment (outsideris) [inline lib/broker.go]: Why is this needed? `dummy` looks like test purpose.

File: lib/broker.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+type EventBroker struct {
+	brokers map[EventType]*Broker
```
Comment (anarcher) [inline lib/broker.go]: IMHO, `EventBroker` seems like  `EventBrokers`.

File: lib/broker.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func NewEventBroker() *EventBroker {
+	e := &EventBroker{
+		brokers: map[EventType]*Broker{},
+	}
+	e.brokers[EventDummy] = NewBroker()
```
Comment (anarcher) [inline lib/broker.go]: IMHO, This statement `e.brokers[EventDummy] = ..` is not necessary.

File: lib/broker.go (hunk 2 above)
Comment (kfangw) [reply lib/broker.go]: Event types are not yet defined. Hence, I added those two.
Maybe those are changed to practical names

File: lib/broker.go (hunk 4 above)
Comment (kfangw) [reply lib/broker.go]: AddBroker much better 👍

File: lib/broker_test.go (hunk 1 above)
Comment (kfangw) [reply lib/broker_test.go]: yeap. I will remove it.

**Comment by kfangw** [comment]
I am working on another testcase.. the testcase may fail.

**Comment by spikeekips** [comment]
I understand why you invented `EventBroker`. I thought you already deeply considered the other implementation of internal event systems. We can make it easier like https://github.com/GianlucaGuarini/go-observable or https://github.com/ReactiveX/RxGo .

**Comment by cmcm2222** [comment]
Wow. 
Reactive programming can be in golang too?? 
Awesome.

**Comment by kfangw** [comment]
Good to know those projects.
I will adopt one one them or make a similar one. Thanks spike!!

**Comment by kfangw** [comment]
@spikeekips @zzim2x @outsideris @anarcher 
please review again.
At this moment, please ignore broken ci build. The broken caused by CI itself

I changed two point
1. server-sent-event api with go-observerble
2. Adopt mux

After this change, i will add bellows
1. client
2. APIs for each issues

**Comment by spikeekips** [comment]
@kfangw Please ask us after finishing work :) Too often changes, I am confused when I start to review.

**Comment by kfangw** [comment]
Writing is done but there is a CI problem...T^T...
Sorry for the annoying...
I will ask again when the problem is solved

**Comment by spikeekips** [comment]
@kfangw Thanks :)

**Comment by kfangw** [comment]
Lets move to https://github.com/bosnet/sebak/pull/114

**Comment by kfangw** [comment]
CI build failed occurred occasionally.
However, the original HANG problem is not seen anymore.
I am sorry for the annoying until now.

Please review this PR

**Comment by outsideris** [comment]
DId you find what caused the hang? just curious.

**Comment by kfangw** [comment]
circleci ran the build with old source.
When I submit another PR with the same branch, It turns to normal...
I don't know what happened in the CI exactly.

**Comment by kfangw** [comment]
@outsideris 
refer this https://circleci.com/gh/bosnet/sebak/190
At that time, I removed the testcase "TestStreaming"
but it ran test with old source.

I guess that the CI couldn't fetch latest source code. 

I have trying to reproduce the problem, but I can't...
Unfortunately, I squashed the commit and force pushed... so i lost the commit T^T

File: lib/api.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+			return
+		case message := <-messageChan:
+			fmt.Fprintf(w, "%s\n", message)
+			flusher.Flush()
+		}
+	}
+}
+
+func GetAccountHandler(ctx context.Context) HandlerFunc {
+	storage, ok := ctx.Value("storage").(*sebakstorage.LevelDBBackend)
```
Comment (anarcher) [inline lib/api.go]: IMHO,  Using `LevelDBBackend` with `context.Context.Value` is not suitable in normal cases.

File: lib/api_test.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	storage, err := sebakstorage.NewStorage(storageConfig)
+	if err != nil {
+		t.Errorf("failed to initialize file db: %v", err)
+	}
+	defer storage.Close()
+
+	ctx := context.WithValue(context.Background(), "storage", storage)
+	router := mux.NewRouter()
+	router.HandleFunc("/account/{address}", GetAccountHandler(ctx)).Methods("GET")
+	server := &http.Server{Addr: ":5000", Handler: router}
```
Comment (anarcher) [inline lib/api_test.go]: IHMO, Using `net/http/httptest` can be better for using testing :-D.

File: lib/block_account.go [hunk 7]
```diff
@@ -42,6 +43,8 @@ func (b *BlockAccount) String() string {
 	return string(sebakcommon.MustJSONMarshal(b))
 }
 
+var BlockAccountObserver = observable.New()
```
Comment (anarcher) [inline lib/block_account.go]: IHMO, In general, **package level vars** as using  **Package-global objects** can have surprising side effects.  Anyway, This var can be in another place. :)

Comment (kfangw) [reply lib/block_account.go]: I don't like this usage also. I am totally agreed what you worry about.
However, the object should be accessed from everywhere.
So, I will just move it to another separated package.
Is it enough? or is there any alternative solution? @anarcher

File: lib/api.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	"boscoin.io/sebak/lib/error"
+	"boscoin.io/sebak/lib/storage"
+	"github.com/GianlucaGuarini/go-observable"
+	"github.com/gorilla/mux"
+	"boscoin.io/sebak/lib/network"
+)
+
+func AddAPIHandlers(ctx context.Context, t *sebaknetwork.HTTP2Network) {
+
+	t.AddAPIHandler(ctx, "/account/{address}", GetAccountHandler).Methods("GET")
```
Comment (kfangw) [inline lib/api.go]: I think that we may need a bunch of features provided by mux package.
hence, we need `*mux.Route` reference here.

@spikeekips

File: lib/block_account.go (hunk 7 above)
Comment (anarcher) [reply lib/block_account.go]: @kfangw  I understand why it has to exist in pkg-global var. It makes sense :-)

File: lib/api_test.go (hunk 6 above)
Comment (kfangw) [reply lib/api_test.go]: I tried the httptest.
May I change when I add another APIs?

File: lib/api.go (hunk 5 above)
Comment (kfangw) [reply lib/api.go]: Issue what you picked is spread all around of handlers for sebak.
I think we need to talk about this in another issue.

**Comment by kfangw** [comment]
@spikeekips @anarcher May I merge this PR?
remaining issues pointed in this PR are ... 

1. unittest with httptest
2. usage of context

those will be solved in another PR.

File: lib/api_test.go (hunk 6 above)
Comment (anarcher) [reply lib/api_test.go]: Using httptest is optional :-)!

**Comment by anarcher** [comment]
IMHO, unittest with httptest is optional. You can ignore it!! :-D

File: lib/api.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"boscoin.io/sebak/lib/network"
+	"github.com/GianlucaGuarini/go-observable"
+)
+
```
Comment (anarcher) [inline lib/api.go]: I know my approach is ugly. `T_T`
```
⚠ func (nr *NodeRunner) Ready() {
      nr.network.SetContext(nr.ctx)
      nr.network.AddHandler(nr.ctx, AddAPIHandlers(nr.storage))
      nr.network.Ready()
  }
```

```
⚠ func AddAPIHandlers(s *sebakstorage.LevelDBBackend) func(ctx context.Context, t *sebaknetwork.HTTP2Network) {
      fn := func(ctx context.Context, t *sebaknetwork.HTTP2Network) {
          //t.AddAPIHandler(ctx, "/account/{address}", GetAccountHandler).Methods("GET")
          t.AddAPIHandler("/account/{address}", GetAccountHandler(ctx, t, s)).Methods("GET")
      }
      return fn
  }
```

```
func GetAccountHandler(ctx context.Context, t *sebaknetwork.HTTP2Network, s *sebakstorage.LevelDBBackend) http.Handle
  rFunc {
      /*
          storage, ok := ctx.Value("storage").(*sebakstorage.LevelDBBackend)
          if !ok {
              panic(errors.New("storage is missing in context"))
          }
      */

      storage := s
```

```
⚠ func (t *HTTP2Network) AddAPIHandler(pattern string, handlerFunc http.HandlerFunc) (router *mux.Route) {
      //pattern := args[0].(string)
      //handler := args[1].(func(context.Context, *HTTP2Network) HandlerFunc)
      apiRouter := t.routers["api"]
      return apiRouter.HandleFunc(pattern, handlerFunc)
  }
```

File: lib/network/http2_network.go [hunk 10]
```diff
@@ -61,15 +69,22 @@ func NewHTTP2Network(config HTTP2NetworkConfig) (h2n *HTTP2Network) {
...
 		server:         server,
+		router:         baseRouter,
 		tlsCertFile:    config.TLSCertFile,
 		tlsKeyFile:     config.TLSKeyFile,
 		receiveChannel: make(chan Message),
 	}
+	h2n.handlers = map[string]func(http.ResponseWriter, *http.Request){}
+	h2n.routers = map[string]*mux.Router{
+		"node": baseRouter.PathPrefix("/node").Subrouter(),
+		"api":  baseRouter.PathPrefix("/api").Subrouter(),
```
Comment (spikeekips) [inline lib/network/http2_network.go]: We can use `UrlPrefixXX` in this place even in map key.

File: lib/network/http2_network.go [hunk 11]
```diff
@@ -10,25 +10,33 @@ import (
...
 	"golang.org/x/net/http2"
 )
 
 type Handlers map[string]func(http.ResponseWriter, *http.Request)
 
+const (
+	UrlNodePrefix = "/node"
+	UrlApiPrefix  = "/api"
+)
+
```
Comment (spikeekips) [inline lib/network/http2_network.go]: In the below code for `HTTP2Network.routers`, you can define url path prefix and router name too.
```go
const (
	RouterNameNode = "node"
	RouterNameAPI  = "node"
)

var (
	UrlPathPrefixNode = fmt.Sprintf("/%s", RouterNameNode)
	UrlPathPrefixAPI  = fmt.Sprintf("/%s", RouterNameAPI)
)
```

**Comment by spikeekips** [comment]
@kfangw I think this PR is the 1st implementation of API using **server side stream**, so how about change the PR title for the original issue, #3 .

File: lib/api.go (hunk 9 above)
Comment (kfangw) [reply lib/api.go]: It is not ugly. elegance. 👍 
I have been suffered from circular import, hence my dirty code been written.
Your idea makes me free from circular import and refer to the storage
Thanks @anarcher

File: lib/network/http2_network.go (hunk 10 above)
Comment (kfangw) [reply lib/network/http2_network.go]: Yeap. I changed it to named const.

File: lib/api.go (hunk 9 above)
Comment (anarcher) [reply lib/api.go]: @kfangw  I am honored that I have helped you. `:-]`
